### PR TITLE
Make the single argument constructor inherit the env

### PR DESCRIPTION
**Pull request by mjcarroll**
Now there are three ways to spawn a subprocess:

* `Subprocess({"env"})` - Spawn subprocess and inherit environment from current executable
* `Subprocess({"env"}, {})` - Spawn subprocess with empty custom environment
* `Subprocess({"env"}, {"FOO=BAR"})` - Spawn subprocess with custom environment.

Opening for comment, still needs tests + docs

### CODE REVIEW

**Comment by azeey** [comment]
Windows CI is not happy

**Comment by mjcarroll** [comment]
> Windows CI is not happy

And not in an expected place, I will take a look.

**Comment by mjcarroll** [comment]
@azeey should be good now

File: include/gz/utils/Subprocess.hh [hunk 1]
```diff
@@ -32,46 +36,80 @@ namespace gz
...
+  /// This variant will spawn a subprocess that uses the user-specified
+  /// environment
+  ///
+  /// \param[in] _commandLine set of arguments starting with an executable
+  ///   used to spawn the subprocess
+  /// \param[in] _environment environment variables to set in the spawned
+  ///   subprocess
   public: Subprocess(const std::vector<std::string> &_commandLine,
-                     const std::vector<std::string> &_environment = {}):
+                     gz::utils::EnvironmentMap _environment):
```
Comment (azeey) [inline include/gz/utils/Subprocess.hh:67]: I like this signature, but would it does break API. Would it be possible to add `Subprocess(const std::vector<std::string> &_commandLine,  const std::vector<std::string> &_environment`)` that would turn around and call this?

Comment (mjcarroll) [reply include/gz/utils/Subprocess.hh:67]: Done, ended up refactoring out my strings -> map -> strings functions so I can use the delegating constructor here.

One hitch is that it makes the map and vector constructors ambiguous:

```
// Could be either
Subprocess({}, {})
```

So now you have to explicitly do:

```
Subprocess({}, gz::utils::EnvironmentMap{});
```

Not a huge deal, but a little less clean imo.

File: include/gz/utils/Subprocess.hh [hunk 2]
```diff
@@ -32,46 +36,96 @@ namespace gz
...
       commandLineCstr.push_back(val.c_str());
     }
     commandLineCstr.push_back(nullptr);
 
-    std::vector<const char*> environmentCstr;
-    for (const auto &val : this->environment)
+    if (!this->inheritEnvironment)
     {
-      environmentCstr.push_back(val.c_str());
+      for (const auto &[key, val] : this->environment)
```
Comment (azeey) [inline include/gz/utils/Subprocess.hh:110]: nit: This can use the new map -> vector function.

**Review by azeey — APPROVED** [review]
LGTM!

File: include/gz/utils/Subprocess.hh (hunk 2 above)
Comment (mjcarroll) [reply include/gz/utils/Subprocess.hh]: Good call, done in a81031c

**Pull request merged**